### PR TITLE
[SPARK-25809][K8S][TEST] New K8S integration testing backends

### PR DESCRIPTION
**Pull request by rvesse**
## What changes were proposed in this pull request?

Currently K8S integration tests are hardcoded to use a `minikube` based backend.  `minikube` is VM based so can be resource hungry and also doesn't cope well with certain networking setups (for example using Cisco AnyConnect software VPN `minikube` is unusable as it detects its own IP incorrectly).

This PR Adds a new K8S integration testing backend that allows for using the Kubernetes support in [Docker for Desktop](https://blog.docker.com/2018/07/kubernetes-is-now-available-in-docker-desktop-stable-channel/).  It also generalises the framework to be able to run the integration tests against an arbitrary Kubernetes cluster.

To Do:

- [x] General Kubernetes cluster backend
- [x] Documentation on Kubernetes integration testing
- [x] Testing of general K8S backend
- [x] Check whether change from timestamps being `Time` to `String` in Fabric 8 upgrade needs additional fix up

## How was this patch tested?

Ran integration tests with Docker for Desktop and all passed:

screen shot 2018-10-23 at 14 19 56

Suggested Reviewers: @ifilonenko @srowen

### CODE REVIEW

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala
Comment (rvesse) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala]: Would people be opposed to bumping the K8S client to 3.1.0 (or even a more recent 3.1.x release) as this would simplify the code significantly and avoid any need to shell out to `kubectl`?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala
Comment (srowen) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala]: (needs an Apache copyright header)

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala
Comment (srowen) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala]: Nit: should these be private?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala
Comment (srowen) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala]: Not sure, but do you want to do the superclass cleanup afterwards, after you finish cleanup in this class?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala
Comment (srowen) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala]: Total nit, but you can write `"kubectl" +: args`

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala
Comment (srowen) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala]: Is it OK that this is basically a singleton? it has state. I'm just wondering if that could cause a problem if used across many tests.

**Review by srowen — COMMENTED** [review]
I don't have substantive comments, just minor ones. Could this be why the k8s tests failed on a fresh Ubuntu install? I didn't do any k8s setup (should I?)

**Comment by rvesse** [comment]
@srowen Yes there are a lot of assumptions made by the integration tests that are not documented anywhere and I figured out by digging in the code and POMs.

Broadly speaking right now to run integration tests you need the following:

- A runnable distribution of Spark built
- `minikube` installed, configured with at least 4 CPUs and 8 GB memory and started

Probably the next thing I will do tomorrow on this is to improve the documentation.  There is a `README.md` in the integration testing module but it omits various details I have encountered working with these recently.

**Comment by skonto** [comment]
@rvesse this capability was in the old integration tests repo [here](https://github.com/apache-spark-on-k8s/spark-integration/blob/master/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/CloudTestBackend.scala).

In our[ fork](https://github.com/lightbend/spark/tree/spark-k8s-2.4-snapshot/resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend) we have merged  that part back, can we add that too? I find it really useful to test real platforms like Openshift etc.

**Comment by skonto** [comment]
@rvesse also the following memory limit works for me:

`minikube config set memory 4096` so maybe g its too much.

**Comment by rvesse** [comment]
@skonto Yep, I plan to do that tomorrow

At least for my `minikube` instance I found 4g insufficient and a couple of tests would fail because their pods didn't get scheduled.  8g is probably over allocating, somewhere in-between likely works also

**Comment by skonto** [comment]
@rvesse I use no vm:
`sudo -E ./minikube start --vm-driver=none`

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala
Comment (ifilonenko) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala]: NIT: line too long, split

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala
Comment (ifilonenko) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala]: Can you add a README to explain how to run with docker-for-desktop, by specifying --deploy-mode

**Review by ifilonenko — COMMENTED** [review]
Thank you so much for this change. This looks pretty good for me, pending small nits and README. 
I tried this change out and it works like a charm. I will have @mccheah and @liyinan926 look at this for merging upon their LGTM. 

Can we make this PR only target docker-for-desktop as I personally would like this to go in. Could a generalization come in a followup? Would like to hear the opinion of the committers

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala
Comment (liyinan926) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala]: No, this is a welcomed. If it's not a significant change to upgrade to 3.1.0, we probably should make it part of this PR.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/ProcessUtils.scala
Comment (liyinan926) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/ProcessUtils.scala]: Should add a `\n` before the first line of the error output.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala]: Fixed

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/docker/DockerForDesktopBackend.scala]: With the latest commits the only state maintained is the K8S client which is reusable and there is no longer any need to modify the on-disk state via `kubectl` by adopting a newer Fabric 8 client library version

**Comment by rvesse** [comment]
@ifilonenko I have done the generalisation today since it was fairly trivial and it actually resolves a number of concerns about the first pass implementation

@skonto I have restored the cloud backend as requested, I ended up having two separate cloud backends.  One using the original code that simply takes a master URL and another that takes a desired K8S context and auto-configures itself from the K8S config file and the context specified.

**Comment by rvesse** [comment]
Ran successfully against one of our dev K8S clusters today:

screen shot 2018-10-25 at 17 39 40

Will continue onto testing against one of our larger production clusters

File: resource-managers/kubernetes/integration-tests/scripts/setup-integration-test-env.sh
Comment (liyinan926) [inline resource-managers/kubernetes/integration-tests/scripts/setup-integration-test-env.sh]: The indention from this point seems to be off.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala
Comment (liyinan926) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: Is this the same as https://github.com/apache/spark/blob/a2f502cf53b6b00af7cb80b6f38e64cf46367595/core/src/main/scala/org/apache/spark/util/Utils.scala#L2725?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala
Comment (liyinan926) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala]: Can we combine `KubeConfigBackend` and `CloudTestBackend` into one? The only difference is how the kubeconfig gets constructed. I guess you can decide which way to go based on the presence of the `CONFIG_KEY_KUBE_MASTER_URL` property.

File: resource-managers/kubernetes/integration-tests/README.md
Comment (mccheah) [inline resource-managers/kubernetes/integration-tests/README.md]: `By default this is set to minikube. The available backends and their prerequisites are described below.`

File: resource-managers/kubernetes/integration-tests/README.md
Comment (mccheah) [inline resource-managers/kubernetes/integration-tests/README.md]: `If you prefer to run just the integration tests directly, then you can customize the behaviour via passing system properties to Maven. For example:`

**Comment by liyinan926** [comment]
@rvesse I want to take back my comment on bumping the version of the fabric8 client in the same PR. Based on the discussion so far, it seems more appropriate to do that in a separate PR, like in #22820. Can you revert the changes to bump the version? Sorry about that.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala
Comment (mccheah) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala]: Just use a default value for `context` instead of checking for null below.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala
Comment (mccheah) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala]: Setting system properties is a little worrisome, for example what if something else sets this system property as well? Would much prefer to treat system properties as immutable after the JVM starts.

**Review by mccheah — CHANGES_REQUESTED** [review]

**Comment by rvesse** [comment]
@liyinan926 I will rebase and squash appropriately once PR #22820 is merged

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala]: Looked over what Fabric 8 client was doing behind the scenes and realised the values I was setting explicitly are the defaults anyway so removed this unnecessary manipulation

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala]: Refactored to avoid this

File: resource-managers/kubernetes/integration-tests/README.md
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/README.md]: Fixed

File: resource-managers/kubernetes/integration-tests/README.md
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/README.md]: Fixed

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala]: Yes, done this in the latest commit

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: Yes, will fix in next commit

File: resource-managers/kubernetes/integration-tests/scripts/setup-integration-test-env.sh
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/scripts/setup-integration-test-env.sh]: Now corrected

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala
Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala]: Now fixed

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala
Comment (liyinan926) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/cloud/KubeConfigBackend.scala]: What if the input `context` is null, e.g., if `CONFIG_KEY_KUBE_CONFIG_CONTEXT` is not defined?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala
Comment (liyinan926) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala]: What if `CONFIG_KEY_KUBE_CONFIG_CONTEXT` is not defined?

Comment (rvesse) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala]: The Fabric 8 client treats `null` as meaning use the users current context from their `KUBECONFIG` file - https://github.com/fabric8io/kubernetes-client/blob/v4.1.0/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java#L467

Comment (liyinan926) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/IntegrationTestBackend.scala]: Got it.

**Review by liyinan926 — COMMENTED** [review]
LGTM. Let us know when you finish your testing.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala
Comment (rvesse) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala]: Note to reviewers.  In testing this I noticed one issue is that when you have multiple contexts defined in your `KUBECONFIG` file that while this PR adds the ability for the integration test machinery to use an arbitrary context the submission client will always use your current context.  Therefore if your current context does not match the requested context the integration tests will fail/return false positives because the client and the test harness will use different contexts.

I have filed SPARK-25887 for addressing this since the client change seemed like it should be a separate issue and PR.  I will provide a PR for that tomorrow.

**Comment by rvesse** [comment]
Did a bunch more testing on our internal K8S clusters today after rebasing this onto master. I am now happy that this is ready for final review and merging so I have removed the `[WIP]` tag

**Comment by ifilonenko** [comment]
test this please

**Review by liyinan926 — APPROVED** [review]
LGTM. Thanks!

**Review by ifilonenko — APPROVED** [review]
LGTM (pending successful tests), thank you for this!

**Comment by ifilonenko** [comment]
retest this please

**Comment by liyinan926** [comment]
@mccheah since he had some comments.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala
Comment (mccheah) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala]: Can we add a Spark ticket here?

**Review by mccheah — COMMENTED** [review]
Just want to make sure we're tracking one follow-up item explicitly, once we add the comment change we can merge. Don't need to specifically wait for CI before the next merge either since it should just be a code comment change.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala
Comment (rvesse) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala]: SPARK-25887, I'll update the comment to reference it explicitly

**Comment by rvesse** [comment]
@mccheah I have updated the comment to reference the follow up issue and opened a PR for that as #22904.  Can we go ahead and merge now?

**Comment by mccheah** [comment]
Yes, this can be merged. Going into master now.

**Pull request closed without merging**
